In [69]:
import pandas as pd
import numpy as np
import time

1. Write a lambda function to rearrange positive and negative numbers in a given list

  Sample output:

  original_list = [-1,2,-3,4,-5,6,-7,8,-9,10,-11,12] 

  resulting_list = [2,4,6,8,10,12,-11,-9,-7,-5,-3,-1]

In [78]:
Result_list= sorted(original_list, key = lambda i: 0 if i == 0 else -1 / i)
Result_list

[2, 4, 6, 8, 10, 12, -11, -9, -7, -5, -3, -1]

2. Create a new dataframe using week_11/temperatures/temperature.csv file from last class. It should have the datetime column and one of the cities. Create a new boolean column called is_freezing that indicates if the temperature is at or below freezing. Use a lambda to create this column. Is there a better way to do this than with a lambda?


In [61]:
df = pd.read_csv("../Week_11/temperature.csv",parse_dates=True, index_col='datetime')
df.head()

Vancouver    Portland  San Francisco     Seattle  \
datetime                                                                 
2012-10-01 12:00:00         NaN         NaN            NaN         NaN   
2012-10-01 13:00:00  284.630000  282.080000     289.480000  281.800000   
2012-10-01 14:00:00  284.629041  282.083252     289.474993  281.797217   
2012-10-01 15:00:00  284.626998  282.091866     289.460618  281.789833   
2012-10-01 16:00:00  284.624955  282.100481     289.446243  281.782449   

                     Los Angeles   San Diego   Las Vegas     Phoenix  \
datetime                                                               
2012-10-01 12:00:00          NaN         NaN         NaN         NaN   
2012-10-01 13:00:00   291.870000  291.530000  293.410000  296.600000   
2012-10-01 14:00:00   291.868185  291.533501  293.403141  296.608509   
2012-10-01 15:00:00   291.862844  291.543355  293.392177  296.631487   
2012-10-01 16:00:00   291.857503  291.553209  293.381213  296.654466   

                     Albuquerque      Denver  ...  Philadelphia    New York  \
datetime                                      ...                             
2012-10-01 12:00:00          NaN         NaN  ...           NaN         NaN   
2012-10-01 13:00:00   285.120000  284.610000  ...    285.630000  288.220000   
2012-10-01 14:00:00   285.154558  284.607305  ...    285.663208  288.247676   
2012-10-01 15:00:00   285.233952  284.599918  ...    285.756824  288.326940   
2012-10-01 16:00:00   285.313345  284.592530  ...    285.850440  288.406203   

                       Montreal      Boston   Beersheba  Tel Aviv District  \
datetime                                                                     
2012-10-01 12:00:00         NaN         NaN         NaN                NaN   
2012-10-01 13:00:00  285.830000  287.170000  307.590000         305.470000   
2012-10-01 14:00:00  285.834650  287.186092  307.590000         304.310000   
2012-10-01 15:00:00  285.847789  287.231672  307.391513         304.281841   
2012-10-01 16:00:00  285.860929  287.277251  307.145200         304.238015   

                          Eilat  Haifa  Nahariyya  Jerusalem  
datetime                                                      
2012-10-01 12:00:00  309.100000    NaN        NaN        NaN  
2012-10-01 13:00:00  310.580000  304.4      304.4      303.5  
2012-10-01 14:00:00  310.495769  304.4      304.4      303.5  
2012-10-01 15:00:00  310.411539  304.4      304.4      303.5  
2012-10-01 16:00:00  310.327308  304.4      304.4      303.5  

[5 rows x 36 columns]

In [79]:
New_df = df.iloc[0:, :1]
New_df

Vancouver
datetime                       
2012-10-01 12:00:00         NaN
2012-10-01 13:00:00  284.630000
2012-10-01 14:00:00  284.629041
2012-10-01 15:00:00  284.626998
2012-10-01 16:00:00  284.624955
...                         ...
2017-11-29 20:00:00         NaN
2017-11-29 21:00:00         NaN
2017-11-29 22:00:00         NaN
2017-11-29 23:00:00         NaN
2017-11-30 00:00:00         NaN

[45253 rows x 1 columns]

In [87]:
start = time.time()

New_df = New_df.assign(is_freezing=lambda x: New_df['Vancouver'] <= 273)
print(New_df)

end = time.time()
print(end-start)

                      Vancouver  is_freezing
datetime                                    
2012-10-01 12:00:00         NaN        False
2012-10-01 13:00:00  284.630000        False
2012-10-01 14:00:00  284.629041        False
2012-10-01 15:00:00  284.626998        False
2012-10-01 16:00:00  284.624955        False
...                         ...          ...
2017-11-29 20:00:00         NaN        False
2017-11-29 21:00:00         NaN        False
2017-11-29 22:00:00         NaN        False
2017-11-29 23:00:00         NaN        False
2017-11-30 00:00:00         NaN        False

[45253 rows x 2 columns]
0.02928614616394043


In [88]:
# created a empty list to hold the result 
start = time.time()

freezing_list = []
# Wrote for loop to check if the temperature for city Vancouver is at or below freezing and added the result to empty list 
for temp in New_df['Vancouver']:
    if temp <= 273:
        freezing_list.append(True)
    else:
        freezing_list.append(False)     
# created column called is_freezing and set that column is equal to freezing list which will convert the list data to coulmns with in the dataframe 
New_df["is_freezing"] = freezing_list
print(New_df)

end = time.time()
print(end-start)

                      Vancouver  is_freezing
datetime                                    
2012-10-01 12:00:00         NaN        False
2012-10-01 13:00:00  284.630000        False
2012-10-01 14:00:00  284.629041        False
2012-10-01 15:00:00  284.626998        False
2012-10-01 16:00:00  284.624955        False
...                         ...          ...
2017-11-29 20:00:00         NaN        False
2017-11-29 21:00:00         NaN        False
2017-11-29 22:00:00         NaN        False
2017-11-29 23:00:00         NaN        False
2017-11-30 00:00:00         NaN        False

[45253 rows x 2 columns]
0.05292797088623047


 ### after comparing lambda and for loop for cities freezing point, i found that lambda function ran more quicker than the for loop. as lambda took 0.029 in time and for loop look 0.052 time. Lambda took less time to run the code then the for loop. It is better to use lambda i think.